connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,431 @@
+
+  var connector: CassandraConnector = _
+  var consolidatedConf: SparkConf = _
+  var catalogOptions: CaseInsensitiveStringMap = _
+  //Something to distinguish this Catalog from others with different hosts
+  var catalogName: String = _
+  var nameIdentifier: String = _
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    catalogOptions = options
+    val sparkConf = sparkSession.sparkContext.getConf
+    consolidatedConf = consolidateConfs(sparkConf, sparkSession.conf.getAll, name, userOptions = options.asScala.toMap)
+    connector = CassandraConnector(consolidatedConf)
+    catalogName = name
+    nameIdentifier = connector.conf.contactInfo.endPointStr()
+  }
+
+  override def name(): String = s"Catalog $catalogName For Cassandra Cluster At $nameIdentifier " //TODO add identifier here
+
+  override def listNamespaces(): Array[Array[String]] = {
+    getMetadata(connector)
+      .getKeyspaces
+      .asScala
+      .map { case (name, _) => Array(name.asInternal()) }
+      .toArray
+  }
+
+  /**
+    * Since we only allow single depth keyspace identifiers in C*
+    * we always either return an empty list of namespaces or
+    * throw a NoSuchNamespaceException
+    */
+  override def listNamespaces(namespace: Array[String]): Array[Array[String]] = {
+    getKeyspaceMeta(connector, namespace) // Thorows no such namespace if namespace is not found
+    Array.empty[Array[String]]
+  }
+
+
+
+  override def createNamespace(namespace: Array[String], metadata: java.util.Map[String, String]): Unit = {
+    val ksMeta = metadata.asScala
+    checkNamespace(namespace)
+    if (getMetadata(connector).getKeyspace(fromInternal(namespace.head)).asScala.isDefined) throw new NamespaceAlreadyExistsException(s"${namespace.head} already exists")
+    connector.withSessionDo { session =>
+      val createStmt = SchemaBuilder.createKeyspace(namespace.head)
+      val replicationClass = ksMeta.getOrElse(ReplicationClass, throw new CassandraCatalogException(s"Creating a keyspace requires a $ReplicationClass DBOption for the replication strategy class"))
+      val createWithReplication = replicationClass.toLowerCase(Locale.ROOT) match {
+        case SimpleStrategy =>
+          val replicationFactor = ksMeta.getOrElse(ReplicationFactor,
+            throw new CassandraCatalogException(s"Need a $ReplicationFactor option with SimpleStrategy"))
+          createStmt.withSimpleStrategy(replicationFactor.toInt)
+        case NetworkTopologyStrategy =>
+          val datacenters = (ksMeta -- IgnoredReplicationOptions).map(pair => (pair._1, pair._2.toInt: java.lang.Integer))
+          createStmt.withNetworkTopologyStrategy(datacenters.asJava)
+        case other => throw new CassandraCatalogException(s"Unknown keyspace replication strategy $other")
+      }
+      val finalCreateStmt = createWithReplication.withDurableWrites(ksMeta.getOrElse(DurableWrites, "True").toBoolean)
+      session.execute(finalCreateStmt.asCql())
```

Comment:
In the alter logic below the CQL statement is built up before we go into withConnectorDo(), effectively minimizing the amount of work we have to do when we have access to the session.  It's not a huge thing but we could do the same here since session isn't actually used until we need to execute the CQL.  I suppose that's probably _slightly_ preferred... but it's not a major thing either way.

Reply:
yeah I think perf wise it probably doesn't matter, but that does sound cleaner